In [1]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

In [9]:
#Read the data
df=pd.read_csv("news.csv")
print(df)

      Unnamed: 0                                              title  \
0           8476                       You Can Smell Hillary’s Fear   
1          10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2           3608        Kerry to go to Paris in gesture of sympathy   
3          10142  Bernie supporters on Twitter erupt in anger ag...   
4            875   The Battle of New York: Why This Primary Matters   
...          ...                                                ...   
6330        4490  State Department says it can't find emails fro...   
6331        8062  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...   
6332        8622  Anti-Trump Protesters Are Tools of the Oligarc...   
6333        4021  In Ethiopia, Obama seeks progress on peace, se...   
6334        4330  Jeb Bush Is Suddenly Attacking Trump. Here's W...   

                                                   text label  
0     Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1     Google Pinter

In [10]:
#Get shape and head
df.shape
df.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [11]:
#DataFlair - Get the labels
labels=df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

In [13]:
#Split the dataset into training and testing sets.
#DataFlair - Split the dataset
x_train,x_test,y_train,y_test=train_test_split(df['text'], labels, test_size=0.2, random_state=7)

**Initializing a TF-IDF Vectorizer** with **stop words from the English language** and a **maximum document frequency of 0.7 **(terms with a higher document frequency will be discarded).

In [14]:
#DataFlair - Initialize a TfidfVectorizer
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)

#DataFlair - Fit and transform train set, transform test set
tfidf_train=tfidf_vectorizer.fit_transform(x_train)
tfidf_test=tfidf_vectorizer.transform(x_test)

Next, we'll **initialize a PassiveAggressiveClassifier**. We'll fit this on **tfidf_train and y_train**.

Then, we'll predict on the test set from the TfidfVectorizer and calculate the accuracy with **accuracy_score()** from sklearn.metrics.

In [20]:
#DataFlair - Initialize a PassiveAggressiveClassifier
pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,y_train)

#DataFlair - Predict on the test set and calculate accuracy
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 92.98%


We got an **accuracy of 92.98%** with this model.

Now finally, let's print out a **confusion matrix** to gain insight into the **number of false and true negatives and positives**.

In [24]:
#DataFlair - Build confusion matrix
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[589,  49],
       [ 40, 589]])

So with this model, we have **589 true positives**, **589 true negatives**,**40 false positives**, and **49 false negatives**. Thus, this means that:

**True Positives (TP):** The number of instances correctly predicted as "FAKE"
(fake news). And this model **successfully detected** **589 fake news** articles.

**True Negatives (TN):** The number of instances correctly predicted as "REAL" (real news). This model also **correctly identified 589 real news **articles.

**False Positives (FP):** The number of instances incorrectly predicted as "FAKE" when they are actually "REAL." These are also known as Type I errors or false alarms. And here, **40 news articles are there where the model predicted as "FAKE," but they are actually "REAL."** This indicates instances where the model made a mistake by classifying real news as fake (Type I error).

**False Negatives (FN):** The number of instances incorrectly predicted as "REAL" when they are actually "FAKE." These are also known as Type II errors or misses. Lastly, **49 news articles are there where the model predicted as "REAL," but they are actually "FAKE."** This represents instances where the model failed to detect fake news (Type II error).